src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,551 @@
+
+        dest_file = os.path.join(upa_dir, assembly_name)
+        loader_helper.create_hardlink_between_files(dest_file, src_file)
+        _update_yaml_file(root_dir, env, workspace_id, upa, assembly_dir_name, assembly_name)
+
+    return upas
+
+
+def upload_assemblies_to_workspace(
+        conf: Conf,
+        workspace_name: str,
+        data_dir: str,
+) -> list[str]:
+    """
+    Upload assemblies to workspace and record failed assembly names.
+
+    conf: Conf object
+    workspace_name: a string used as a name for a workspace
+    data_dir: directory of assemblies to upload
+    """
+
+    assembly_files = os.listdir(data_dir)
+    print(f'start uploading {len(assembly_files)} assembly files')
+
+    failed_names = list()
+
+    counter = 1
+    for assembly_name in assembly_files:
+
+        if counter % 5000 == 0:
+            print(f"{round(counter / len(assembly_files), 4) * 100}% finished at {datetime.now()}")
+
+        assembly_path = os.path.join(JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER, DATA_DIR, assembly_name)
+        try:
+            _upload_assembly_to_workspace(conf, workspace_name, assembly_path, assembly_name)
+        except Exception as e:
+            print(e)
+            failed_names.append(assembly_name)
+
+        counter += 1
+
+    if failed_names:
+        print(f'Failed to upload {failed_names}')
+
+    return failed_names
+
+
+def main():
+    parser = _get_parser()
+    args = parser.parse_args()
+
+    workspace_id = args.workspace_id
+    kbase_collection = getattr(args, loader_common_names.KBASE_COLLECTION_ARG_NAME)
+    source_version = getattr(args, loader_common_names.SOURCE_VER_ARG_NAME)
+    root_dir = args.root_dir
+    token_filepath = args.token_filepath
+    upload_file_ext = args.upload_file_ext
+    keep_job_dir = args.keep_job_dir
+
+    env = args.env
+    kb_base_url = loader_common_names.KB_BASE_URL_MAP[env]
+
+    if workspace_id <= 0:
+        parser.error(f"workspace_id needs to be > 0")
+
+    uid = os.getuid()
+    username = os.getlogin()
+
+    job_dir = loader_helper.make_job_dir(root_dir, username)
+    collection_source_dir = loader_helper.make_collection_source_dir(
+        root_dir, loader_common_names.DEFAULT_ENV, kbase_collection, source_version
+    )
+    upload_dir = loader_helper.make_collection_source_dir(root_dir, env, kbase_collection, source_version)
+    output_dir = loader_helper.make_sourcedata_ws_dir(root_dir, env, workspace_id)
+
+    proc = None
+    conf = None
+
+    try:
+        # start podman service
+        proc = loader_helper.start_podman_service(uid)
+    except Exception as e:
+        print(e)
+    else:
+        # set up conf, start callback server, and upload assemblies to workspace
+        conf = Conf(job_dir, kb_base_url, token_filepath)
+        workspace_name = conf.ws.get_workspace_info({"id": workspace_id})[1]
+
+        count, wait_to_upload_assemblies = _fetch_assemblies_to_upload(
+            root_dir,
+            env, 
```

Comment:
Shouldn't this be DEFAULT_ENV? We're loading file from there into a workspace `env` right?

Reply:
should be `env`? DEFAULT_ENV is `NONE`

It is into workspace, and it will check what assemblies already live in that workspace through YAML file.

Reply:
No, because at the beginning of the process, the directories in `collectionssource/GTDB/None` will be linked into `sourcedir`, and so can see the yaml files, but the ones in `collectionssource/GTDB/CI` (for example) won't be linked.

Reply:
DEFAULT_ENV info is already in collection_source_dir when pass into _fetch_assemblies_to_upload function. 

To be clear, we probably need two separate `env` here:
1. env for loading data from collectionssource
2. env for uploading to workspace

Reply:
There are 2 different envs:
* The env created by the gtdb/ncbi downloader. This is always None and is the source of the data.
* The env provided when running the uploader, which is one of the 4 KBase envs. The data is added here

Reply:
But I see where I was confused. `env` is passed into _fetch_assemblies, which passes it to _read_yaml_file. However, it is *not* used to find the yaml file path, just when looking at the contents of the file since they're keyed by `env`.

Reply:
Maybe there's a different variable name we could give the variable to prevent that confusion. Maybe `upload_env_key` or something?

Reply:
replaced `env` by `upload_env_key`.

Reply:
:+1: